experimental/casemapping/src/provider/mod.rs

```diff
@@ -57,16 +68,40 @@ pub mod unfold;
 /// convert characters and strings to upper, lower, or title case.
 pub struct CaseMappingV1<'data> {
     /// Case mapping data
-    #[cfg_attr(feature = "serde", serde(borrow))]
     pub trie: CodePointTrie<'data, CaseMappingData>,
-    #[cfg_attr(feature = "serde", serde(borrow))]
     /// Exceptions to the case mapping data
     pub exceptions: CaseMappingExceptions<'data>,
-    #[cfg_attr(feature = "serde", serde(borrow))]
     /// Reverse case folding data.
     pub unfold: CaseMappingUnfoldData<'data>,
 }
 
+#[cfg(feature = "serde")]
+impl<'de> serde::Deserialize<'de> for CaseMappingV1<'de> {
+    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        #[derive(serde::Deserialize)]
+        pub struct Raw<'data> {
+            #[serde(borrow)]
+            pub trie: CodePointTrie<'data, CaseMappingData>,
+            #[serde(borrow)]
+            pub exceptions: CaseMappingExceptions<'data>,
+            #[serde(borrow)]
+            pub unfold: CaseMappingUnfoldData<'data>,
+        }
+
+        let raw = Raw::deserialize(deserializer)?;
+        let result = Self {
+            trie: raw.trie,
+            exceptions: raw.exceptions,
+            unfold: raw.unfold,
+        };
+        use serde::de::Error;
+        result
```

Comment:
issue: this should be in debug mode only

(I know there are a couple `unwraps()` that rely on `validate()` but I'm getting rid of them)

Reply:
Actually, I would be down to this being a debug assertion in the constructors, even, so that we can still rely on the serde derive.

Reply:
I'm only moving the `validate` in this PR, from the constructor to deserialize. We can make changes in another PR

Reply:
Yep, and it's a debug assertion in the constructor :)

Reply:
fixed